hpx/parallel/executors/persistent_auto_chunk_size.hpp

```diff
@@ -0,0 +1,126 @@
+#include <hpx/parallel/executors/executor_information_traits.hpp>
+#include <hpx/util/high_resolution_clock.hpp>
+#include <hpx/util/date_time_chrono.hpp>
+
+#include <cstddef>
+#include <algorithm>
+
+#include <boost/cstdint.hpp>
+
+namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v3)
+{
+    ///////////////////////////////////////////////////////////////////////////
+    /// Loop iterations are divided into pieces and then assigned to threads.
+    /// The number of loop iterations combined is determined based on
+    /// measurements of how long the execution of 1% of the overall number of
+    /// iterations takes.
+    /// This executor parameters type makes sure that as many loop iterations
+    /// are combined as necessary to run for the amount of time specified.
+    ///
+    struct persistent_auto_chunk_size : executor_parameters_tag
+    {
+    public:
+        /// Construct an \a persistent_auto_chunk_size executor parameters object
+        ///
+        /// \note Default constructed \a persistent_auto_chunk_size executor parameter
+        ///       types will use 80 microseconds as the minimal time for which
+        ///       any of the scheduled chunks should run.
+        ///
+
+
+        /// Construct an \a persistent_auto_chunk_size executor parameters object
+        ///
+        /// \param rel_time     [in] The time duration to use as the minimum
+        ///                     to decide how many loop iterations should be
+        ///                     combined.
+        ///
+
+        explicit persistent_auto_chunk_size(hpx::util::steady_duration time_cs = boost::chrono::nanoseconds(0))
+          : chunk_size_time_(time_cs.value().count())
+        {}
+
+        /// \cond NOINTERNAL
+        // Estimate a chunk size based on number of cores used.
+        template <typename Executor, typename F>
+        std::size_t get_chunk_size(Executor& exec, F && f, std::size_t count)
+        {
+            std::size_t const cores = executor_information_traits<Executor>::
+                processing_units_count(exec, *this);
+
+            if (count > 100*cores)
+            {
+                boost::uint64_t t;
+                if(chunk_size_time_ == 0)
+                {
+                    using hpx::util::high_resolution_clock;
+                    t = high_resolution_clock::now();
+
+                    std::size_t test_chunk_size = f();
+                    if (test_chunk_size != 0)
+                    {
+                        
+                        t = (high_resolution_clock::now() - t) / test_chunk_size;
+                        chunk_size_time_ = t;
+                    
+                        if (t != 0)
+                        {
+                            // return chunk size which will create the required
+                            // amount of work
+                            return (std::min)(count, (std::size_t)(min_time_ / t));
+                        }
```

Comment:
Sure, the issue is that the parallel algorithm which calls this function will use the returned value to fix the number of items it executes in the same chunk, it will not interpret this number in terms of time.